scripts/util/estimate-twin-transfer-limits.js

```diff
@@ -0,0 +1,210 @@
+var log = console.log;
+
+// Modify the console.log to capture the contract output
+var contractLogs = [];
+console.log = function () {
+  contractLogs.push(arguments[0]);
+};
+
+const hre = require("hardhat");
+
+/**
+Estimates the values for SINGLE_TWIN_RESERVED_GAS, MINIMAL_RESIDUAL_GAS
+
+Use with caution if `transferTwins` or other methods that call it changed a lot. Please, refer to docs/twin-transfer-limits.md for more information.
+
+This script does the following:
+1. Preprocesses the `transferTwins` method to add the `gasLeft()` measurement points
+   If it cannot reliably position the measurement points, it throws an error
+2. Runs the unit tests that cover the `transferTwins` method
+3. Captures the console.log output
+4. Analyzes the output and estimates SINGLE_TWIN_RESERVED_GAS and MINIMAL_RESIDUAL_GAS
+
+*/
+async function estimateTwinTransferLimits() {
+  // 1. Preprocess the `twinTransfer` to add the `gasLeft()` measurement points
+  let invokeFound = false;
+  let loopFound = false;
+  let externalCallFound = false;
+  let paranthesisCount = {
+    invoke: 0,
+    loop: 0,
+    externalCall: 0,
+  };
+  let previousMeasurePoint = "";
+  const ExchangeHandlerPath = process.cwd() + "/contracts/protocol/facets/ExchangeHandlerFacet.sol";
+  hre.config.preprocess = {
+    eachLine: () => ({
+      transform: (line, { absolutePath }) => {
+        // Filter paths here, since preprocessor's options "files" doesn't work
+        if (absolutePath !== ExchangeHandlerPath) {
+          return line;
+        }
+
+        // 1. add console.sol import
+        if (line.includes("pragma solidity")) {
+          return line + "import 'hardhat/console.sol';";
+        }
+
+        // Gas left measure points
+        // 1. At the beginning of the transfer twin loop
+        if (line.includes("for (uint256 i = 0; i < twinCount")) {
+          if (previousMeasurePoint !== "invoke") {
+            throw new Error("Could not find the invoke measure point");
+          }
+          previousMeasurePoint = "loopStart";
+          loopFound = true;
+          return line + 'console.log("TwinGasEstimate1,%s,%s", i, gasleft());';
```

Comment:
```suggestion
          return line + 'console.log(`TwinGasEstimate1,${i},${gasleft()}`);';
```

(similar suggestion for the others console.log)

Reply:
This part of the script preprocesses `.sol` files and solidity does not support template literals or string interpolation, so compilation fails.
We need to stick with the existing formatting.